[dynamicIO] Disallow only dynamic metadata (#78576)
vercel/next.js: errors/next-prerender-dynamic-metadata.mdx

---
title: Cannot access Request information or uncached data in `generateMetadata()` in an otherwise entirely static route
---

## Why This Error Occurred

> When `dynamicIO` is enabled, Next.js requires that `generateMetadata()` not depend on uncached data or Request data unless unless some other part of the page also has similar requirements. The reason for this is that while you normally control your intention for what is allowed to be dynamic by adding or removing Suspense boundaries in your Layout and Page components you are not in control of rendering metadata itself.

Reviewer: ```suggestion
When `dynamicIO` is enabled, Next.js requires that `generateMetadata()` not depend on uncached data or Request data unless some other part of the page also has similar requirements. The reason for this is that while you normally control your intention for what is allowed to be dynamic by adding or removing Suspense boundaries in your Layout and Page components, you are not in control of rendering metadata itself.
```

> The heuristic Next.js uses to understand your intent with `generateMetadata()` is to look at the data requirements of the rest of the route. If other components have dependence on Request data and uncached data then we allow `generateMetadata()` to have similar data requirements. If the rest of your page has no dependence on this type of data we require that `generateMetadata()` also not have this type of data dependence.

Reviewer: ```suggestion
The heuristic Next.js uses to understand your intent with `generateMetadata()` is to look at the data requirements of the rest of the route. If other components depend on Request data or uncached data, then we allow `generateMetadata()` to have similar data requirements. If the rest of your page has no dependence on this type of data, we require that `generateMetadata()` also not have this type of data dependence.
```

## Possible Ways to Fix It

To fix this issue, you must first determine your goal for the affected route.

### Caching External Data

> If your metadata does not depend on any request data then it may be possible for you to indicate the data is cacheable which would allow Next.js to include it in the prerender for this route. Consider using `"use cache"` to mark the function producing the external data as cacheable.

Reviewer: ```suggestion
If your metadata does not depend on any request data, then it may be possible for you to indicate that the data is cacheable, which would allow Next.js to include it in the prerender for this route. Consider using `"use cache"` to mark the function producing the external data as cacheable.
```

  // Next.js will interpret this as needing to be rendered
  // on every request.
  const { title } = await cms.getPageData('/.../page')
  return {
    title,
  }
}

async function getPageText() {
  'use cache'
  const { text } = await cms.getPageData('/.../page')
  return text
}

export default async function Page() {
  // This text is cached so the main content of this route
  // is prerenderable.
  const text = await getPageText()
  return <article>{text}</article>
}
```

After:

```jsx filename="app/.../page.tsx"
import { cms } from './cms'

export async function generateMetadata() {
  // By marking this function as cacheable, Next.js
  // can now include it in the prerender for this route.
  'use cache'
  const { title } = await cms.getPageData('/.../page')
  return {
    title,
  }
}

async function getPageText() {
  'use cache'
  const { text } = await cms.getPageData('/.../page')
  return text
}

export default async function Page() {
  // This text is cached so the main content of this route
  // is prerenderable.
  const text = await getPageText()
  return <article>{text}</article>
}
```

### If you must access Request Data or your external data is uncacheable

> If your metadata requires request specific data or depends on external data which is not cacheable then Next.js will need to render this page dynamically on every request. However if you got this error the rest of your page is able to be completely static. This is generally pretty rare but if this is your actual constraint you can indicate to Next.js that the page should be allowed to be dynamic by rendering any other component that is dynamic. Since your route doesn't have any genuine dynamic requirements you might render an artificially dynamic component using `await connection()`. This is like telling Next.js that this component is never prerenderable and must be rendered on every user request.

Reviewer: ```suggestion
If your metadata requires request-specific data or depends on external data which is not cacheable, then Next.js will need to render this page dynamically on every request. However, if you got this error, the rest of your page is able to be completely static. This is generally pretty rare, but if this is your actual constraint, you can indicate to Next.js that the page should be allowed to be dynamic by rendering any other component that is dynamic. Since your route doesn't have any genuine dynamic requirements, you might render an artificially dynamic component using `await connection()`. This is like telling Next.js that this component is never prerenderable and must be rendered on every user request.
```

> If your metadata requires request specific data or depends on external data which is not cacheable then Next.js will need to render this page dynamically on every request. However if you got this error the rest of your page is able to be completely static. This is generally pretty rare but if this is your actual constraint you can indicate to Next.js that the page should be allowed to be dynamic by rendering any other component that is dynamic. Since your route doesn't have any genuine dynamic requirements you might render an artificially dynamic component using `await connection()`. This is like telling Next.js that this component is never prerenderable and must be rendered on every user request.

Reviewer: ```suggestion
If your metadata requires request-specific data or depends on external data which is not cacheable, then Next.js will need to render this page dynamically on every request. However, if you got this error, the rest of your page is able to be completely static. This is generally pretty rare, but if this is your actual constraint, you can indicate to Next.js that the page should be allowed to be dynamic by rendering any other component that is dynamic. Since your route doesn't have any genuine dynamic requirements, you might render an artificially dynamic component using `await connection()`. This is like telling Next.js that this component is never prerenderable and must be rendered on every user request.
```

Before:

```jsx filename="app/.../page.tsx"
import { cookies } from 'next/headers'

> export async function generateViewport() {

Reviewer: Probably copy & paste from `next-prerender-dynamic-viewport.mdx`. Should be a `generateMetadata` snippet, right?

  const cookieJar = await cookies()
  return {
    themeColor: cookieJar.get('theme-color'),
  }
}

export default function Page() {
  return <main>This page is entirely static</main>
}
```

After:

```jsx filename="app/.../page.tsx"
import { Suspense } from 'react'
import { cookies } from 'next/headers'
import { connection } from 'next/server'

export async function generateViewport() {
  const cookieJar = await cookies()
  return {
    themeColor: cookieJar.get('theme-color'),
  }
}

async function DynamicMarker() {
  // This component renders nothing but it will always
  // be dynamic because it waits for an actual connection

>   const NullDynamic = async () => { await connection(): return null }

Reviewer: ```suggestion
  const NullDynamic = async () => { await connection(); return null }
```

  return <Suspense><NullDynamic /></Suspense>
}

export default function Page() {
  return (
    <>

>       <main>This page is entirely static</main>

Reviewer: Should this text be updated now that the page is not entirely static anymore after adding `<DynamicMarker />`?

      <DynamicMarker />
    </>
  )
}
```

> Note: The reason to structure this DynamicMarker as a self-contained Suspense boundary is to avoid blocking the actual content of the page from being prerendered. When Partial Prerendering is enabled alongside dynamicIO the static shell will still contain all of the prerenderable content still and only the metadata will stream in dynamically.

Reviewer: ```suggestion
Note: The reason to structure this `DynamicMarker` as a self-contained Suspense boundary is to avoid blocking the actual content of the page from being prerendered. When Partial Prerendering is enabled alongside `dynamicIO`, the static shell will still contain all of the prerenderable content, and only the metadata will stream in dynamically.
```

## Useful Links

> - [`generateMetadata()`](docs/app/api-reference/functions/generate-metadata)
> - [`connection()`](docs/app/api-reference/functions/connection)
> - [`cookies()`](docs/app/api-reference/functions/cookies)

Reviewer: ```suggestion
- [`generateMetadata()`](/docs/app/api-reference/functions/generate-metadata)
- [`connection()`](/docs/app/api-reference/functions/connection)
- [`cookies()`](/docs/app/api-reference/functions/cookies)
```

This mistake was also done in `next-prerender-dynamic-viewport.mdx`.

  return text
}

export default async function Page() {
  // This text is cached so the main content of this route
  // is prerenderable.
  const text = await getPageText()
  return <article>{text}</article>
}
```

### If you must access Request Data or your external data is uncacheable

If your metadata requires Request data or depends on external data which is not cacheable then Next.js will need to render this page dynamically on every request. However if you got this error, the rest of your page is able to be completely static. This is generally pretty rare, but if this is your actual constraint, you can indicate to Next.js that the page should be allowed to be dynamic by rendering any other component that is dynamic. Since your route doesn't have any genuine need for Request data, you can indicate an intentional dependency on a Request with `await connection()`. This is like telling Next.js that this component is never prerenderable and must be rendered on every user request.

Before:

```jsx filename="app/.../page.tsx"
import { cookies } from 'next/headers'
import { getPersonalizedTitle } from './my-api'

export async function generateMetadata() {

>   // In this example we are assuming we must fetch our title
>   // from an protected external API. While the response is potentially
>   // cacheable the it still requires accessing a token from the Request cookies.

Reviewer: ```suggestion
  // In this example, we are assuming we must fetch our title
  // from a protected external API. While the response is potentially
  // cacheable, it still requires accessing a token from the Request cookies.
```

  const token = (await cookies()).get('token')
  const response = await getPersonalizedTitle(token)
  return {
    title: getTitle(response),
  }
}

export default function Page() {
  return <article>This article is completely static</article>
}
```

After:

```jsx filename="app/.../page.tsx"
import { Suspense } from 'react'
import { cookies } from 'next/headers'
import { connection } from 'next/server'

export async function generateMetadata() {
  const token = (await cookies()).get('token')
  const response = await fetch(..., { headers: { Authorization: token } })
  return {
    title: getTitle(response),
  }
}

async function DynamicMarker() {

>   // This component renders nothing but it will always
>   // be dynamic because it waits for an actual connection

Reviewer: ```suggestion
  // This component renders nothing, but it will always
  // be dynamic because it waits for an actual connection.
```
